### ISSUE TITLE
Booting from a partition is broken in multiple ways.

### ISSUE THREAD

**Issue by circl-lastname** [SerenityOS/serenity#10898]
Sorry for combining issues, but these are very interlocked. I have tried to run SerenityOS on VirtualBox, and have found many bugs related to partitions.
## Cannot boot a partition with `root=/dev/hda1`.
https://github.com/SerenityOS/serenity/blob/f19cb6bb223c58035762eddfb8b2b596cd9eac9f/Kernel/Storage/StorageManagement.cpp#L123
This function while supposed to get all block devices, including entire disks and partitions, only finds disks, as `m_storage_devices` only contains disks. Trying to boot with `root=/dev/hda1` results in a kernel panic, but this is used by grub after doing `ninja grub-image`.
## Cannot boot with `root=PARTUUID=<uuid>` either.
This is because of this code, which prevents two equal signs in a boot argument.
https://github.com/SerenityOS/serenity/blob/f19cb6bb223c58035762eddfb8b2b596cd9eac9f/Kernel/CommandLine.cpp#L70
## You cannot generate a GPT formatted image to actually have UUIDs present.
Doing `ninja grub-image gpt` doesn't work, as it asks me if I meant "ps". Modifying ninja.build to pass `gpt` to `build-image-grub.sh` results in an out of space error.
## Bonus problem: Partitions don't show up even after booting.
With a second disk with no partition table, and the other with grub, I saw that there was no `/dev/hda1` even after booting! Only `/dev/hda`.
## The fix.
While all of these are problems, I think just fixing booting with `root=/dev/hda1` will be enough for now. As for `PARTUUID`, simply changing the prefix and making a larger disk image will work.

**Comment by unixinspace** [comment]
I did a little digging here and found that you can get the same result trying to boot the grub-image in qemu. 

I threw in some dbgln's and determined that in StorageManagement::determine_boot_device() on the standard _disk_image,
storage_device.early_storage_name and storage_name are both set to hda, as the _disk_image doesn't have a partition table, it's a raw ext2 filesystem. Changing this allows us to pass the check at 
```
        for (auto& storage_device : m_storage_devices) {
            dbgln("DLS: issue? esn: {}, sn: {}", storage_device.early_storage_name(), storage_name);
            if (storage_device.early_storage_name() == storage_name) {
                m_boot_block_device = storage_device;
            }
        }

```

I tried booting as you did above and by changing the kernel parameter to /dev/hda, the ext2 driver to complain that we have bad magic on our superblock.
```
0.998 [#0 init_stage2(1:1)]: DLS: issue? esn: hda, sn: hda
1.026 [init_stage2(1:1)]: Ext2FS: Bad super block magic
[init_stage2(1:1)]: KERNEL PANIC! :^(
[init_stage2(1:1)]: StorageManagement: Couldn't open root filesystem: Error(errno=22)
[init_stage2(1:1)]: at ../../Kernel/Storage/StorageManagement.cpp:196 in AK::NonnullRefPtr<Kernel::FileSystem> Kernel::StorageManagement::root_filesystem() const
[init_stage2(1:1)]: Kernel + 0x0082e78f  Kernel::__panic(char const*, unsigned int, char const*) +0x115
[init_stage2(1:1)]: Kernel + 0x0058edd9  Kernel::StorageManagement::root_filesystem() const +0x573
[init_stage2(1:1)]: Kernel + 0x009f6364  Kernel::init_stage2(void*) +0x148b
[init_stage2(1:1)]: Kernel + 0x00a096d7  exit_kernel_thread +0x0
qemu-system-i386: terminating on signal 2

```

So now to check the magic being passed, and it appears it is not.
`1.118 [init_stage2(1:1)]: Ext2FS: super block magic: 0000 (super block size: 1024)`

Alas this really reaches the end of my ability to debug for now. I'll look more into it later, but hopefully this helps someone look deeper into the issue.

**Comment by circl-lastname** [comment]
The reason you cannot do /dev/hda to boot is because it refers to the entire disk, and trying to interpret as ext2 results in an error. Like I said in my issue, you cannot boot using `/dev/hda1` because the partion isn't in `m_storage-devices`. This might be realted to the partition not working after booting, suggesting they're just broken.

**Comment by circl-lastname** [comment]
Okay, I have found that every disk in `m_storage_devices` has `.partitions()`, which is used here.
https://github.com/SerenityOS/serenity/blob/f19cb6bb223c58035762eddfb8b2b596cd9eac9f/Kernel/Storage/StorageManagement.cpp#L151
I think it is as simple as adding a second for loop to `determine_boot_device` using this .

**Comment by circl-lastname** [comment]
> Okay, I have found that every disk in `m_storage_devices` has `.partitions()`, which is used here.
> 
> https://github.com/SerenityOS/serenity/blob/f19cb6bb223c58035762eddfb8b2b596cd9eac9f/Kernel/Storage/StorageManagement.cpp#L151
> 
> I think it is as simple as adding a second for loop to `determine_boot_device` using this .

There might not be a nice function to actually get the name of the partition.

**Comment by circl-lastname** [comment]
Closed: fixed by ed5a4f2
